instrumentation/kafka/kafka-clients/kafka-clients-common/library/src/main/java/io/opentelemetry/instrumentation/kafka/internal/ConsumerAndRecord.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.kafka.internal;
+
+import com.google.auto.value.AutoValue;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import javax.annotation.Nullable;
+import org.apache.kafka.clients.consumer.Consumer;
+
+/**
+ * This class is internal and is hence not for public use. Its APIs are unstable and can change at
+ * any time.
+ */
+@AutoValue
+public abstract class ConsumerAndRecord<R> {
+
+  public static <R> ConsumerAndRecord<R> create(@Nullable Consumer<?, ?> consumer, R record) {
```

Comment:
I also tried implementing the same metrics. Instead of consumer + record I went with record + consumerGroup + clientId. I suspect it will be hard to get access to consumer/producer in `TracingConsumerInterceptor` and `TracingProducerInterceptor` while the string values can be extracted in `configure` method.
Another potential issue is that `ConsumerAndRecord` could potentially be injected in multiple class loaders if for example kafka client is in parent and spring is in child. We have soled similar issues by moving such shared classes to bootstrap, but then we'd have library module depending on a bootstrap module which I think we don't have elsewhere.